docs/cvl/hooks.md

```diff
@@ -20,77 +22,422 @@ Syntax
+production.
+
+(load-hooks)=
+(store-hooks)=
+Load and store hooks
+--------------------
+
+Load hooks are executed before a read from a specific location in storage, while
+store hooks are executed before a write to a specific location in storage.
+
+The locations to be matched are given by an access path, such as a contract
+variable, array index, or a slot number.  See {ref}`access-paths` below for
+information on the available access paths.
+
+A load pattern contains the keyword `Sload`, followed by the type and name of a
+variable that will hold the loaded value, followed by an access path indicating
+the location that is read.  Load patterns must end with the keyword `STORAGE`.
+
+For example, here is a load hook that will execute whenever a contract reads the
+value of `C.owner`:
+```cvl
+hook Sload address o C.owner STORAGE { ... }
+```
+Inside the body of this hook, the variable `o` will be bound to the value that
+was read.
+
+A store pattern contains the keyword `Sstore`, followed by an access path
+indicating the location that is being written to, followed by the type and name
+of a variable to hold the value that is being stored.  Optionally, the pattern
+may also include the type and name of a variable to store the previous value
+that is being overwritten.  Store patterns must end with the keyword `STORAGE`.
+
+For example, here is a store hook that will execute whenever a contract writes
+the value of `C.totalSupply`:
+```cvl
+hook Sstore C.totalSupply uint ts (uint old_ts) STORAGE { ... }
+```
+Inside the body of this hook, the variable `ts` will be bound to the value that
+is being written to the `totalSupply` variable, while `old_ts` is bound to the
+value that was stored there previously.
+
+If you do not need to refer to the old value, you can omit the variable
+declaration for it.  For example, the following hook only binds the new value
+of `C.totalSupply`:
+```cvl
+hook Sstore C.totalSupply uint ts STORAGE { ... }
+```
+
+```{todo}
+Is this correct?
+
+If there is a store hook that binds the old value of the variable, then the
+Prover will add a load instruction that reads the value immediately before the
+store instruction.  Therefore, if a path has both a load hook and a store hook,
+they will both be executed when the contract performs a store.
+```
+
+(access-paths)=
+### Access paths
+
+The patterns for load and store hooks are fine-grained; they allow you to hook
+on accesses to specific contract variables or specific array, struct, or
+mapping accesses.
+
+Storage locations are designated by "access paths".  An access path
+starts with either the name of a contract field, or a [slot number][storage-layout].
+
+```{todo}
+is this correct?
+
+Contract fields must be qualified by the contract that defines them (e.g.
+`Contract.field`).  If the contract name is omitted, it defaults to
+`currentContract`.
+```
+
+```{todo}
+Does the contract need to be the contract that defines the field, or can it be
```

Comment:
iirc we don't care about the inheritance for naming/identifying fields. if the same private field appears in more than one contract in the inheritance structure, we omit those completely and they are inaccessible. the classic example is the `_gap` static array used by upgradable contracts